### DISCUSSION TITLE
V1 -> V2 advice and or best practice for ios_format(apn)?

### DISCUSSION THREAD

**Discussion by rromanchuk** [excid3/noticed#406] [General]
I used to utilize the `recipient` to construct some apn payload properties and additional branching and now it's out of scope, so thinking about the correct way to re-implement, without a new custom delivery.

Some examples where I use the recipient:
- Fine tuning `apn.badge` from recipient level counters
- In some contexts, branching during payload construction (vs config.if) to deliver silent content_available=1 or bannered depending on comm preferences. So the channel is always delivering, but the payload construction changes based on preferences, just a layer down

Example of a typical v1 apn

```ruby
def ios_format(apn)
  apn.custom_payload = { notification_id: record.id, universal_link: url, image_url: }

  if recipient.setting_for_this_story.include?('push')
    apn.badge = recipient.notifications_badge_number
    apn.alert = { title: title, body: message}
    apn.mutable_content = 1
    apn.priority = HIGH_PRIORITY
    apn.sound = 'chicken-alarm.caf'
    apn.interruption_level = 'time-sensitive'
    apn.relevance_score = 0.8
  else
    apn.content_available = 1
    apn.priority = LOW_PRIORITY
  end
end
```

**Comment by excid3** [comment]
Some examples would probably help me understand what you're trying to do.

**Reply by rromanchuk** [reply]
Updated above. Thought more about it, I think silent notifications actually makes sense as being it's own custom delivery method. Especially they way i'm trying to use it, as a fallback, which works well with `if`, `unless` config.  For the badge, i'm just thinking about making it a `required_option` on delivery method config.

This is about as complex as it could get and seems a lot less smelly than what i was doing before.  

```ruby
# frozen_string_literal: true

module Apnable
  extend ActiveSupport::Concern

  included do
    # User subscribed/enabled/allowed via preferences
    deliver_by :ios do |config|
      config.format = :ios_format
      config.apns_key = Credentials[:apn_key]
      config.key_id = Credentials[:apn_key_id]
      config.team_id = Credentials[:apn_team_id]
      config.bundle_identifier = Rails.configuration.general.ios_bundle_id
      config.device_tokens = :ios_device_tokens
      config.if = :push_notifications?
    end
    
     # A silent apn that may still be important with content=1 to refresh local state  
    deliver_by :ios_silient, class: "Noticed::DeliveryMethods::Ios" do |config|
      config.format = :ios_silient_format
      config.apns_key = Credentials[:apn_key]
      config.key_id = Credentials[:apn_key_id]
      config.team_id = Credentials[:apn_team_id]
      config.bundle_identifier = Rails.configuration.general.ios_bundle_id
      config.device_tokens = :ios_device_tokens
      config.unless = :push_notifications?
    end
     
     # A tiny tiny use case when i'm testing n-environments from xcode builds
    deliver_by :ios_development, class: "Noticed::DeliveryMethods::Ios" do |config|
      config.format = :ios_format
      config.apns_key = Credentials[:apn_key]
      config.key_id = Credentials[:apn_key_id]
      config.team_id = Credentials[:apn_team_id]
      config.bundle_identifier = Rails.configuration.general.ios_bundle_id
      config.development = true
      config.device_tokens = :ios_development_device_tokens
      config.if = :push_notifications?
    end
  end

  LOW_PRIORITY = 5
  HIGH_PRIORITY = 10

  # def ios_cert
  #   StringIO.new(Credentials[:apn_key])
  # end

  def ios_device_tokens(notification)
    notification.recipient.apple_tokens.where.not(environment: :development).pluck(:token)
  end

  def ios_development_device_tokens(notification)
    notification.recipient.apple_tokens.where(environment: :development).pluck(:token)
  end

  def push_notifications?(notification)
    raise NotImplementedError
  end

  def invalid_token(token)
    AppleToken.where(token:).destroy_all
  end
end
```

**Reply by excid3** [reply]
This looks like it separates them out logically which was what I was hoping we could achieve.

**Reply by rromanchuk** [reply]
OK last question promise, this is obviously anti-pattern, and injecting `deliver_by :db`, so let me give context.

I have a `GET /api/notifications/:id`  that returns a `Noticed::Notification.find params[:id]`, well, more like a`current_user.notifications.find params[:id]`, but you get it,  which allows the relational persistent, client side store, to prefetch what a view will likely be displaying. 

My custom payloads only include `{universal_link: https://foo.com/record/:id, notification_id:}`, where the universal link is used for navigation, and notification id is used to offload complexity at the data/network layer. 

What's the best way to inject the AR `Noticed::Notification` PK into my apn payload? I tried `id` in scope, but it's the `Noticed::Event`, or `event_id`, which is fine, but makes it a bit more clumsy routing, and i want to avoid even exposing those details. 

Rationale
- Limit apns from being used as a client side data layer (privacy/DRY/etc)
- Remove complexity in `userNotificationCenter:center:willPresent` et al, becomes a single fetch, without knowledge of record id or type, where delegation is two lines `try await NotificationRequest.show(notificationId).save()` -> `UL(url: universalLink).handle()`

**Reply by excid3** [reply]
If you use a Proc/lambda, it will evaluate in the notification context.

We probably should update the Symbol option to pass in the notification after the apn.

https://github.com/excid3/noticed/blob/main/lib/noticed/delivery_methods/ios.rb#L37-L41

**Reply by excid3** [reply]
@rromanchuk I pushed a commit to `main` (meant to do it in a branch) but that should fix your issue. It was always intended to pass the notification and the apn to both procs and methods.
